### PR TITLE
Add Multi-Fractional Non-Fungible Token Standard

### PR DESCRIPTION
**Pull request by PowerStream3604**
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

File: EIPS/eip-###.md [hunk 1]
```diff
@@ -0,0 +1,224 @@
+---
+eip: <to be assigned>
```
Comment (MicahZoltu) [inline EIPS/eip-###.md:2]:
```suggestion
eip: 4675
```
Please also update the file name to `EIPS/eip-4675.md`

File: EIPS/eip-###.md [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+title: Multi-Fractional Non-Fungible Token Standard
+description: 
+author: David Kim(@powerstream3604)
+discussions-to: https://github.com/ethereum/EIPs/issues/4674
```
Comment (MicahZoltu) [inline EIPS/eip-###.md:6]: `discussions-to` should point to a forum thread at https://ethereum-magicians.org/, not a GitHub issue.

Out of curiosity, where did you see that a GitHub issue should be created?  We recently changed the policy on this and if there is some place we forgot to update it would be great to find and fix it!

Comment (PowerStream3604) [reply EIPS/eip-###.md]: I think I was not updated about that.
I created a topic in ethereum magicians and added that
topic link for "discussions to" link.

File: EIPS/eip-4675.md [hunk 3]
```diff
@@ -0,0 +1,224 @@
+---
+eip: 4675
+title: Multi-Fractional Non-Fungible Token Standard
+description:
```
Comment (SamWilsn) [inline EIPS/eip-4675.md:4]: Should have one full short sentence here.

File: EIPS/eip-4675.md [hunk 4]
```diff
@@ -0,0 +1,224 @@
+---
+eip: 4675
+title: Multi-Fractional Non-Fungible Token Standard
+description: 
+author: David Kim(@powerstream3604)
```
Comment (SamWilsn) [inline EIPS/eip-4675.md:5]:
```suggestion
author: David Kim (@powerstream3604)
```

File: EIPS/eip-4675.md [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+discussions-to: https://ethereum-magicians.org/t/eip-4675-multi-fractional-non-fungible-token-standard/8008
+status: Draft
+type: Standards Track
+category: ERC
+created: 2021-01-13
+requires: 165
```
Comment (SamWilsn) [inline EIPS/eip-4675.md:11]: Might also be worth mentioning ERC-721?

File: EIPS/eip-4675.md [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+Also, the sequence of parameter in use of ``_id`` for distinguishing token types in ``functions`` and ``events`` are very much similar to ERC-1155 Multi-Token Standard.
+
+Since this standard is intended to interact with the EIP-721 Non-Fungible Token Standard, it is kept purposefully agnostic to extensions beyond the standard in order to allow specific projects to design their own token usage and scenario.
+
+## Test Cases
+
+Reference Implementation of MFNFT Token includes test cases written using hardhat. (Test coverage : 100%)
+
+## References
```
Comment (wschwab) [inline EIPS/eip-4675.md:200]: There isn't a `References` section in [EIP-1](./eip-1.md) - please remove this section. In terms of the information, you can put it in a different section if you deem it necessary, but please use relative links, so instead of:
```
[ERC-20](https://eips.ethereum.org/EIPS/eip-20)
```
use:
```
[ERC-20](./erc-20.md)
```

File: EIPS/eip-4675.md [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+Also, the sequence of parameter in use of `_id` for distinguishing token types in `functions` and `events` are very much similar to ERC-1155 Multi-Token Standard.
+
+Since this standard is intended to interact with the EIP-721 Non-Fungible Token Standard, it is kept purposefully agnostic to extensions beyond the standard in order to allow specific projects to design their own token usage and scenario.
+
+## Test Cases
+
+Reference Implementation of MFNFT Token includes test cases written using hardhat. (Test coverage : 100%)
+
+## Reference Implementation
+[MFNFT - Implementation](https://github.com/PowerStream3604/multi-fnft)
```
Comment (SamWilsn) [inline EIPS/eip-4675.md:201]: EIP-1 was recently updated to recommend against external links (Sorry!)

Would it be possible to include the reference implementation in the `assets` directory, as described [here](https://eips.ethereum.org/EIPS/eip-1#auxiliary-files)?

File: EIPS/eip-4675.md [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+To prevent this issue, implementors should just allow **admin** to call, or fractionalize and receive NFT in an atomic transaction similar to flash loan(swap).
+
+## Copyright
+Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).
+
+## Citation
+
+Please cite this document as:
+
+[David Kim](mailto:helloyongjun3604@gmail.com) "EIP-4675: Multi-Fractional Non-Fungible Token"
```
Comment (Pandapip1) [inline EIPS/eip-4675.md:218]:
```suggestion
Copyright and related rights waived via [CC0](../LICENSE.md).
```

File: EIPS/eip-4675.md (hunk 7 above)
Comment (PowerStream3604) [reply EIPS/eip-4675.md]: I appreciate your update! @SamWilsn 
Deleted all external links and pointed it to ../assets

**Comment by PowerStream3604** [comment]
@lightclient, @axic, @SamWilsn I request to check if this PR is eligible to be merged.

File: EIPS/eip-4675.md [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+To fractionalize an already minted NFT, it is evident that ownership of NFT should be given to token contracts before fractionalization.
+In the case of fractionalizing NFT, the token contract should thoroughly verify the ownership of NFT before fractionalizing it to prevent tokens from being a separate tokens with the NFT.
+
+If an arbitrary account has the right to call `setParentNFT()` there might be a front-running issue. The caller of `setParentNFT()` might be different from the real NFT sender. 
+To prevent this issue, implementors should just allow **admin** to call, or fractionalize and receive NFT in an atomic transaction similar to flash loan(swap).
+
+## Copyright
+Copyright and related rights waived via [CC0](../LICENSE.md).
+
+## Citation
```
Comment (lightclient) [inline EIPS/eip-4675.md:214]: Please remove this section, it is added automatically on the website `eips.ethereum.org`.

**Review by lightclient — CHANGES_REQUESTED** [review]

File: EIPS/eip-4675.md [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+type: Standards Track
+category: ERC
+created: 2021-01-13
+requires: 165
+---
+
+## Abstract
+This standard outlines a smart contract interface eligible to represent any number of fractionalized non-fungible tokens. Existing projects utilizing standards like EIP-1633 conventionally deploy seperate ERC-20 compatible token standard to fractionalize the non-fungible token into ERC-20 tokens. In contrast, this ERC allows each token ID to represent a token type representing(fractionalizing) the non-fungible token.
+
+This standard is approximate in terms of using ``_id`` for distinguishing token types. However, this ERC has a clear difference with ERC-1155 as each ``_id`` represents a distinct NFT.
```
[the remark this answers is not in the document]
Comment (PowerStream3604) [reply EIPS/eip-4675.md]: This is already updated in the current version.
The current version uses a single back tick to denote code.

File: EIPS/eip-4675.md (hunk 9 above)
Comment (lightclient) [reply EIPS/eip-4675.md:211]: @PowerStream3604, this comment is on the "Citation" section. The "Copyright" section is required.

**Review by lightclient — CHANGES_REQUESTED** [review]

File: EIPS/eip-4675.md (hunk 9 above)
Comment (PowerStream3604) [reply EIPS/eip-4675.md]: @lightclient , I updated the file as requested.
Sorry, I misunderstood it.

File: EIPS/eip-4675.md [hunk 11]
```diff
@@ -0,0 +1,212 @@
+---
+eip: 4675
+title: Multi-Fractional Non-Fungible Token Standard
+description: Token standard to handle multiple NFTs in a fractionalized form
```
Comment (SamWilsn) [inline EIPS/eip-4675.md:4]:
```suggestion
title: Multi-Fractional Non-Fungible Tokens
description: Fractionalize multiple NFTs using a single contract
```

We try to avoid "standard" in titles and descriptions.

File: EIPS/eip-4675.md [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+discussions-to: https://ethereum-magicians.org/t/eip-4675-multi-fractional-non-fungible-token-standard/8008
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-01-13
+requires: 165, 721
+---
+
+## Abstract
+This standard outlines a smart contract interface eligible to represent any number of fractionalized non-fungible tokens. Existing projects utilizing standards like EIP-1633 conventionally deploy seperate ERC-20 compatible token standard to fractionalize the non-fungible token into ERC-20 tokens. In contrast, this ERC allows each token ID to represent a token type representing(fractionalizing) the non-fungible token.
```
Comment (SamWilsn) [inline EIPS/eip-4675.md:15]: The first mention of another EIP must be linked, use EIP instead of ERC, and I think you had the wrong word in here:

```suggestion
This standard outlines a smart contract interface eligible to represent any number of fractionalized non-fungible tokens. Existing projects utilizing standards like [EIP-1633](./eip-1633.md) conventionally deploy separate [EIP-20](./eip-20.md) compatible token contracts to fractionalize the non-fungible token into EIP-20 tokens. In contrast, this ERC allows each token ID to represent a token type representing(fractionalizing) the non-fungible token.
```

File: EIPS/eip-4675.md [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+
+This standard is approximate in terms of using `_id` for distinguishing token types. However, this ERC has a clear difference with ERC-1155 as each `_id` represents a distinct NFT.
```
Comment (SamWilsn) [inline EIPS/eip-4675.md:17]:
```suggestion
This standard is approximate in terms of using `_id` for distinguishing token types. However, this ERC has a clear difference with [EIP-1155](./eip-1155.md) as each `_id` represents a distinct NFT.
```

File: EIPS/eip-4675.md [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+    @return `bytes4(keccak256("onERC721Received(address,address,uint256,bytes)"))`
+*/
+function onERC721Received(address _operator, address _from, uint256 _tokenId, bytes calldata _data) external pure returns (bytes4);
+```
+
+## Rationale
+
+**Metadata**
+
+The `symbol()` & `name()` function was not included since the majority of users can just fetch it from the originating NFT contract. Also, copying the name & symbol every time when token gets added might place a lot of redundant bytecode on the Ethereum blockchain.
```
Comment (SamWilsn) [inline EIPS/eip-4675.md:180]:
```suggestion
The `symbol()` & `name()` functions were not included since the majority of users can just fetch it from the originating NFT contract. Also, copying the name & symbol every time when token gets added might place a lot of redundant bytecode on the Ethereum blockchain. 
```

File: EIPS/eip-4675.md [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+Also, the sequence of parameter in use of `_id` for distinguishing token types in `functions` and `events` are very much similar to ERC-1155 Multi-Token Standard.
+
+Since this standard is intended to interact with the EIP-721 Non-Fungible Token Standard, it is kept purposefully agnostic to extensions beyond the standard in order to allow specific projects to design their own token usage and scenario.
+
+## Test Cases
+
+Reference Implementation of MFNFT Token includes test cases written using hardhat. (Test coverage : 100%)
+
+## Reference Implementation
+[MFNFT - Implementation](../assets/eip-4675/)
```
Comment (SamWilsn) [inline EIPS/eip-4675.md:201]: You might need to link to a specific file (like a README) to get the bot to accept your PR.

File: assets/eip-4675/contracts/math/SafeMath.sol [hunk 16]
```diff
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: ISC
```
Comment (SamWilsn) [inline assets/eip-4675/contracts/math/SafeMath.sol:1]: I am not a lawyer, but if this is copied from OpenZeppelin, I don't think you can just change the license.

File: assets/eip-4675/coverage.json [hunk 17]
```diff
@@ -0,0 +1 @@
...9,"type":"if","locations":[{"start":{"line":339,"column":8},"end":{"line":339,"column":8}},{"start":{"line":339,"column":8},"end":{"line":339,"column":8}}]}}},"contracts/NFT.sol":{"l":{"13":24,"17":0,"24":25,"33":0},"path":"/Users/kimyongjun/DEV/multi-fnft/contracts/NFT.sol","s":{"1":24,"2":0,"3":25,"4":0},"b":{},"f":{"1":17,"2":24,"3":0,"4":25,"5":0},"fnMap":{"1":{"name":"constructor","line":10,"loc":{"start":{"line":10,"column":4},"end":{"line":10,"column":44}}},"2":{"name":"safeMint","line":12,"loc":{"start":{"line":12,"column":4},"end":{"line":14,"column":4}}},"3":{"name":"_baseURI","line":16,"loc":{"start":{"line":16,"column":4},"end":{"line":18,"column":4}}},"4":{"name":"_beforeTokenTransfer","line":20,"loc":{"start":{"line":20,"column":4},"end":{"line":25,"column":4}}},"5":{"name":"supportsInterface","line":27,"loc":{"start":{"line":27,"column":4},"end":{"line":34,"column":4}}}},"statementMap":{"1":{"start":{"line":13,"column":8},"end":{"line":13,"column":29}},"2":{"start":{"line":17,"column":8},"end":{"line":17,"column":21}},"3":{"start":{"line":24,"column":8},"end":{"line":24,"column":52}},"4":{"start":{"line":33,"column":8},"end":{"line":33,"column":51}}},"branchMap":{}}}
```
Comment (SamWilsn) [inline assets/eip-4675/coverage.json:1]: Is this file necessary for the reference implementation?

File: assets/eip-4675/package-lock.json [hunk 18]
```diff
@@ -0,0 +1,18117 @@
+{
```
Comment (SamWilsn) [inline assets/eip-4675/package-lock.json:1]: You'll probably need to delete this file to get the spellcheck to pass.

**Comment by PowerStream3604** [comment]
Thank you for your precise feedback @SamWilsn 
I updated the files as requested.

**Review by SamWilsn — COMMENTED** [review]
Few spelling mistakes left, then it should be good to merge in draft!

```
./EIPS/eip-4675.md:15: seperate ==> separate
./EIPS/eip-4675.md:20: seperated ==> separated
./EIPS/eip-4675.md:41: ot ==> to, of, or
./assets/eip-4675/README.md:5: everytime ==> every time
./assets/eip-4675/test/MFNFT.General.spec.ts:142: trys ==> tries
```

**Comment by PowerStream3604** [comment]
> Few spelling mistakes left, then it should be good to merge in draft!
> 
> ```
> ./EIPS/eip-4675.md:15: seperate ==> separate
> ./EIPS/eip-4675.md:20: seperated ==> separated
> ./EIPS/eip-4675.md:41: ot ==> to, of, or
> ./assets/eip-4675/README.md:5: everytime ==> every time
> ./assets/eip-4675/test/MFNFT.General.spec.ts:142: trys ==> tries
> ```

Fixed as requested! @SamWilsn

**Review by SamWilsn — APPROVED** [review]

File: assets/eip-4675/contracts/NFT.sol [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    function _baseURI() internal pure override returns (string memory) {
+        return "Test";
+    }
+
+    function _beforeTokenTransfer(address from, address to, uint256 tokenId)
+    internal
+    override(ERC721, ERC721Enumerable)
+    {
```
Comment (venkateshpamarthi) [inline assets/eip-4675/contracts/NFT.sol:23]: this function takes 4 arguments; but you have given three

Comment (PowerStream3604) [reply assets/eip-4675/contracts/NFT.sol:23]: It could be three as well; check [here](https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#ERC20-_beforeTokenTransfer-address-address-uint256-)
